test/cql-pytest/test_shedding.py

```diff
@@ -67,8 +67,8 @@ def get_protocol_errors():
         # tries to establish a connection with another node, and throws a NoHostAvailable exception if it fails.
         # In the debug builds, the driver can have time to grab the error from the socket,
         # and we get InvalidRequest exception.
```

Comment:
Why we get `WriteFailure`? The comments above describe the cases when we get `NoHostAvailable` and `InvalidRequest`, but I didn't see `WriteFailure`. I will try it tomorrow.

Reply:
> Why we get WriteFailure?

That's exactly the question :) My guess is that unless it's #8204 strikes back (real regression), it can be either another race (my server is notably much more powerful than CI nodes) or local misconfiguration

Reply:
The commit message suggests that these are "different driver versions".

I'm having a hard time understanding this test at all. The hack to run the requests on a separate "session" (TCP connection) suggests that this feature doesn't even work correctly, because the point of that feature was to break the one request but let the next requests work - and if that doesn't work, what's the point at all?

According to the driver documentation and code, WriteFailure means that the coordinator sent a write to the replica, and that replica write failed. But isn't the point of this feature to recognize the oversized request already on the coordinator, and not send it to the replica? If we do reject oversized requests in the replica, then a WriteFailure can be understandable. In fact it is a more correct version of NoHostAvailable (since a host **is** available - it's just resulted in an error).

Reply:
> I'm having a hard time understanding this test at all. The hack to run the requests on a separate "session" (TCP connection) suggests that this feature doesn't even work correctly, because the point of that feature was to break the one request but let the next requests work - and if that doesn't work, what's the point at all?

There is a comment about it near the test:

> # Big request causes the entire connection to be closed (see #8800 for the reasons),

https://github.com/scylladb/scylladb/pull/8800

> According to the driver documentation and code, WriteFailure means that the coordinator sent a write to the replica, and that replica write failed. But isn't the point of this feature to recognize the oversized request already on the coordinator, and not send it to the replica?

Agreed. Maybe @xemul uses custom Scylla config/cmdline? The test highly depends on the values [here](https://github.com/scylladb/scylladb/blob/master/test/pylib/scylla_cluster.py#L114).

Reply:
run the test on the current master on my dev machine, it passed

Reply:
> Agreed. Maybe @xemul uses custom Scylla config/cmdline? The test highly depends on the values [here](https://github.com/scylladb/scylladb/blob/master/test/pylib/scylla_cluster.py#L114).

@xemul uses `./test.py --mode=dev shedding` :) If it's very sensitive to some configuration, then it should be enforced by the test itself, not rely on my box being configured properly

Reply:
> > Agreed. Maybe @xemul uses custom Scylla config/cmdline? The test highly depends on the values [here](https://github.com/scylladb/scylladb/blob/master/test/pylib/scylla_cluster.py#L114).
> 
> If it's very sensitive to some configuration, then it should be enforced by the test itself, not rely on my box being configured properly

It's not relying on anything except the tests being run though `test.py`.

Reply:
> run the test (`./test.py test_shedding`) on the current master on my dev machine, it passed

It failed on @alecco 's node too

Reply:
> > run the test (`./test.py test_shedding`) on the current master on my dev machine, it passed
> 
> It failed on @alecco 's node too

Yes, we are investigating it.

Reply:
> According to the driver documentation and code, WriteFailure means that the coordinator sent a write to the replica, and that replica write failed.

@nyh , in my case there likely was only one replica. I think so, because in testlog directory there was only one `scylla-%d.log` file

Reply:
Sure, cql-pytest is always a single node, so single replica. There are still different code paths (and errors) that happen in the coordinator and in the replica.

I can commit this patch "because it works" (TM), but I'm a bit worried that maybe we are breaking the testing power of this test (if the original bug this test was testing regresses, will the test still be able to fail?).

Reply:
> I can commit this patch "because it works" (TM)

No, we should figure out what's happening. I'm dealing with it now.

Reply:
> Sure, cql-pytest is always a single node, so single replica. There are still different code paths (and errors) that happen in the coordinator and in the replica.
> 
> I can commit this patch "because it works" (TM), but I'm a bit worried that maybe we are breaking the testing power of this test (if the original bug this test was testing regresses, will the test still be able to fail?).

From my side -- I'm all for _not_ committing it ASAP and investigating the failure root cause for exactly same reason

Reply:
I created a [fix](https://github.com/scylladb/scylladb/pull/12854).

Reply:
I'm afraid I don't understand why that is a fix for the same problem. If compression caused the request to be too small to be labeled "too big", why did the write fail at all?

Reply:
> I'm afraid I don't understand why that is a fix for the same problem. If compression caused the request to be too small to be labeled "too big", why did the write fail at all?

The request is decompressed at the cql transport level, the replica tries to process a 60mb request and hits some internal limits.